CriteoPublisherSdk/Tests/IntegrationTests/CR_DfpRewardedFunctionalTests.m

```diff
@@ -43,26 +46,82 @@ @implementation CR_DfpRewardedFunctionalTests
 
 - (void)setUp {
   // WARNING AdUnitHelper does use singleton object (no DI in here) should be changed with DPP-3734
-  [Criteo.sharedCriteo.dependencyProvider.integrationRegistry declare:CR_IntegrationGamAppBidding];
+  self.oldIntegrationRegistry = Criteo.sharedCriteo.dependencyProvider.integrationRegistry;
 }
 
 - (void)tearDown {
-  [Criteo.sharedCriteo.dependencyProvider.integrationRegistry declare:CR_IntegrationFallback];
+  Criteo.sharedCriteo.dependencyProvider.integrationRegistry = self.oldIntegrationRegistry;
 }
 
-- (void)test_givenRewarded_whenLoad_thenFailToReceiveAd {
+- (void)test_givenRewarded_noIntegrationDeclared_whenLoadThreeTimes_adLoaded {
   CRRewardedAdUnit *adUnit = [CR_TestAdUnits rewarded];
 
   self.criteo = [Criteo testing_criteoWithNetworkCaptor];
+  // use same integrationRegistry in test criteo object and global instance (see DPP-3734)
+  CR_IntegrationRegistry *integrationRegistry = [[CR_IntegrationRegistry alloc] init];
+  self.criteo.dependencyProvider.integrationRegistry = integrationRegistry;
+  Criteo.sharedCriteo.dependencyProvider.integrationRegistry = integrationRegistry;
+
+  // register here fails as integration is not yet detected
   [self.criteo testing_registerWithAdUnits:@[ adUnit ]];
 
   GAMRequest *request = [[GAMRequest alloc] init];
 
+  // this one make the detection logic discover GAM integration
+  [self enrichAdObject:request forAdUnit:adUnit];
+
+  // register is now ok due to GAM being detected
+  [self.criteo testing_registerWithAdUnits:@[ adUnit ]];
+
+  // this one warms up the cache
+  [self enrichAdObject:request forAdUnit:adUnit];
+
+  // this one passes properly
   [self enrichAdObject:request forAdUnit:adUnit];
+
   CR_DependencyProvider *dependencyProvider = self.criteo.dependencyProvider;
+  CR_CdbBid *bid = [dependencyProvider.cacheManager
+      getBidForAdUnit:[CR_AdUnitHelper cacheAdUnitForAdUnit:adUnit]];
+  NSDictionary *targeting = request.customTargeting;
+
+  XCTAssertEqualObjects(targeting[CR_TargetingKey_crtCpm], @"1.12");
+  XCTAssertEqual(targeting[CR_TargetingKey_crtFormat], @"video");
+  XCTAssertEqualObjects(targeting[CR_TargetingKey_crtSize], @"320x480");
 
+  // as this is a "video" url is not base64ed
+  NSString *encodedUrl = targeting[CR_TargetingKey_crtDfpDisplayUrl];
+  NSString *decodedUrl =
+      [[encodedUrl stringByRemovingPercentEncoding] stringByRemovingPercentEncoding];
+
+  XCTAssertEqualObjects(bid.displayUrl, decodedUrl);
+  BOOL finished = [self.criteo testing_waitForRegisterHTTPResponses];
+  XCTAssert(finished, "Failed to receive all prefetch requests");
+}
+
+- (void)test_givenRewarded_GAMIntegrationDeclared_whenLoadTwoTimes_adLoaded {
+  CRRewardedAdUnit *adUnit = [CR_TestAdUnits rewarded];
+
+  self.criteo = [Criteo testing_criteoWithNetworkCaptor];
+  // use same integrationRegistry in test criteo object and global instance (see DPP-3734)
+  CR_IntegrationRegistry *integrationRegistry = [[CR_IntegrationRegistry alloc] init];
```

Comment:
Do you mind extracting this into a méthod. So this Jacky duplication get wrapped

Reply:
Extracted to mark the code needing adjustments.